Implementation for KEP-4831: Control VPA eviction behavior based on scaling direction and resource

voelzmo opened the pull request:
#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
* Implements `ScalingDirectionPodEvictionAdmission`, which controls Pod admission to eviction based on `EvictionRequirements` defined in a VPA's `.spec.updatePolicy.evictionRequirements`
* Makes Updater use the new `ScalingDirectionPodEvictionAdmission` instead of `DefaultPodEvictionAdmission`
* Change e2e tests to use VPA builder that's already used in the unit tests
  * Changes to VPA builder to allow for multiple containers and more detailed settings per container 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Implements #4831
API PR done in #5176 

#### Special notes for your reviewer:
Depends on the API changes in #5176

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added `.updatePolicy.evictionRequirements` to the VPA spec allowing to constrain Pod eviction based on scaling direction (scaling up or scaling down) and resource (cpu or memory).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [AEP]: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/enhancements/4831-control-eviction-behavior
```

voelzmo commented:
@jbartosik Can you please take a look? Thanks!

voelzmo commented:
@jbartosik Can you try to give this a review, please? Thanks!

jbartosik commented on vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission.go:
```diff
@@ -0,0 +1,104 @@
+	}
+	for _, container := range pod.Spec.Containers {
+		recommendedResources := vpa_utils.GetRecommendationForContainer(container.Name, resources)
+		if recommendedResources == nil {
```
This seems incorrect. This will admit pod if at least one container:
- doesn't have a recommendation or
- has a recommendation in the "correct" direction

I think this should return false if it finds at leas 1 container that has recommendation going in the wrong direction, default to true
?

jbartosik replied:
Description of the function[1] and AEP[2] say that all `EvictionRequirement`s must be met to allow eviction. 

If that's not what you need then we can revisit the KEP (at least rename and have the description match implementation; I didn't think to much about if allowing eviction if at least one container matches is better than requiring all of them match restriction).

[1]
> Admit denies Pod eviction, if an EvictionRequirement exists for at least one Container which evaluates to 'false'.

[2]
[permalink](https://github.com/kubernetes/autoscaler/blob/d3ec0c4d049084bc5bb266a19448df0e82e3056a/vertical-pod-autoscaler/enhancements/4831-control-eviction-behavior/README.md?plain=1#L24):

> A single `EvictionRequirement` defines a condition which must be `true` to allow eviction for the corresponding `Pod`. When multiple `EvictionRequirements` are specified for a `Pod`, all of them must evaluate to `true` to allow eviction.

voelzmo replied:
I think what the AEP mentions still holds true:
> A single EvictionRequirement defines a condition which must be true to allow eviction for the corresponding Pod. When multiple EvictionRequirements are specified for a Pod, all of them must evaluate to true to allow eviction.

In [`allowEvictionForResourceAndScalingDirection`](https://github.com/kubernetes/autoscaler/blob/dcec527d41e1604c64609ea6640833242db9f85a/vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission.go#L71) we iterate over all `EvictionRequirements` and if _one_ of the Requirements returns `false`, we return `false` for the entire Pod.

The critical point is _when_ an `EvictionRequirement` is `true` for a `Pod` based on the individual evaluations for the `Containers`.

I realize that the current comment is incorrect, I think I moved this one back and forth a few times
> Admit denies Pod eviction, if an EvictionRequirement exists for at least one Container which evaluates to 'false'.

This probably should be 
> Admit admits a Pod for eviction in one of three cases 
>  * an EvictionRequirement exists which is evaluated to 'true' for at least one Container in this Pod
>  * no Recommendation exists for at least one Container in this Pod
>  * no EvictionRequirement exists for this Pod

Does this make sense? Do we need a clarification in the AEP as well regarding what it means that an `EvictionRequirement` is true for a `Pod`?

jbartosik replied:
Thanks for the explanation. I'll think about this a bit, I hope I'll be back with a proper reply on Monday

jbartosik replied:
My understanding of the AEP:

- If there is only 1 eviction requirement, on 1 resource we don't evict unless one resource changes in the specified direction
- If there is only 1 eviction requirement, on both resources we don't evict unless at least one resource changed 
- If there are 2 eviction requirements, on both resources separately we don't evict unless both resources change in their specified directions

This code treats separate `EvictionRequirement`s  for CPU the way KEP describes putting them in a single `EvictionRequirement` (one of them changes in the specified direction, not necessarily both). This seems wrong to me

Also I think we should clarify in the AEP that "A single EvictionRequirement defines a condition which must be true to allow eviction for the corresponding Pod" means that at least one container has a change in the specified direction. This makes sense but got me confused for a while

voelzmo replied:
Good catch, thanks! I think there was a test case missing that covered the scenario of a single `EvictionRequirement` with multiple resources, where only one of them was matched and the Pod was still admitted.
I added the case and fixed the code to make it green.

The other scenario ensuring that multiple `EvictionRequirements` are connected with `AND` was already in the test cases and still works.

I've also restructured the code a bit to match the comments and early exists for the cases in which we admit a Pod, hope this makes more sense now?

voelzmo replied:
Opened PR https://github.com/kubernetes/autoscaler/pull/5989 to address this part of the comment 

> Also I think we should clarify in the AEP that "A single EvictionRequirement defines a condition which must be true to allow eviction for the corresponding Pod" means that at least one container has a change in the specified direction.

jbartosik commented on vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission.go:
```diff
@@ -0,0 +1,104 @@
+func (s *scalingDirectionPodEvictionAdmission) allowEvictionForResourceAndScalingDirection(resource apiv1.ResourceName, currentRequests int64, recommendation int64, podEvictionRequirements []*vpa_types.EvictionRequirement) bool {
+	for _, evictionRequirement := range podEvictionRequirements {
+		for _, resourceName := range evictionRequirement.Resources {
+			if resourceName == resource {
```
if resourceName != resource {
  continue
}

then we have 1 less indent level

voelzmo commented:
Force-pushed an update:
* rebasing on most recent state of #5176 and adapting the test cases to https://github.com/kubernetes/autoscaler/pull/5896
* removing the `OrEqual` ChangeRequirements and test cases as discussed in https://github.com/kubernetes/autoscaler/pull/5754#issuecomment-1647577540

jbartosik commented on vertical-pod-autoscaler/e2e/v1/common.go:
```diff
@@ -297,14 +297,15 @@ func SetupVPA(f *framework.Framework, cpu string, mode vpa_types.UpdateMode, tar
 
 // SetupVPAForNHamstersWithMinReplicas creates and installs a simple hamster VPA for a pod with n containers, setting MinReplicas. To be used for e2e test purposes.
-func SetupVPAForNHamstersWithMinReplicas(f *framework.Framework, n int, cpu string, mode vpa_types.UpdateMode, targetRef *autoscaling.CrossVersionObjectReference, minReplicas *int32) *vpa_types.VerticalPodAutoscaler {
+func SetupVPAForNHamstersWithMinReplicas(f *framework.Framework, n int, cpu string, mode vpa_types.UpdateMode, targetRef *autoscaling.CrossVersionObjectReference, minReplicas *int32, er []*vpa_types.EvictionRequirement) *vpa_types.VerticalPodAutoscaler {
```
Seems like we're never setting both minReplicas and er; Please:

- Add a new helper

or even better

- Add a builder (which will be cleaner to use)

voelzmo replied:
Thanks, that makes sense! I'll make the builders available in the e2e tests and push an update soon.

voelzmo replied:
done.

jbartosik commented on vertical-pod-autoscaler/e2e/v1/updater.go:
```diff
@@ -60,34 +60,104 @@ var _ = UpdaterE2eDescribe("Updater", func() {
+		ginkgo.By("Waiting for pods to be evicted")
+		err := WaitForPodsEvicted(f, podList)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	})
```
```suggestion
	})

```

jbartosik commented on vertical-pod-autoscaler/e2e/v1/updater.go:
```diff
@@ -60,34 +60,104 @@ var _ = UpdaterE2eDescribe("Updater", func() {
 		err := WaitForPodsEvicted(f, podList)
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 	})
+	ginkgo.It("evicts pods for downscaling", func() {
```
```suggestion

	ginkgo.It("evicts pods for downscaling", func() {
```

jbartosik commented on vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go:
I'm reviewing this in #5176 and presume we'll not change this file in this PR

voelzmo commented:
force-pushed an update:
* rebased on current master that includes the merged API PR
* refactorings to expose VPA builder in e2e tests
* modifications to VPA builder such that it can replace the previous `SetupVPAForNHamstersWithMinReplicas` functions

Open points:
* one TODO in the VPA builder
* Simplifications/inlining for the `SetupVPAForNHamstersWithMinReplicas` functions

I can get to that next week

voelzmo commented:
@jbartosik This is good for another round of review from my side. I've addressed the last remaining comments and switched to using the VPABuilder in the e2e tests.

jbartosik commented on vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission_test.go:
```diff
@@ -0,0 +1,390 @@
+package priority
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
```
This should be with other k8s imports

voelzmo replied:
done.

jbartosik reviewed (commented):
Mostly looks good, I'm leaving some small comments; I want to take another look at vertical-pod-autoscaler/pkg/updater/priority/scaling_direction_pod_eviction_admission_test.go before merging

voelzmo commented on vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go:
```diff
@@ -153,7 +153,7 @@ func (calc *UpdatePriorityCalculator) GetSortedPods(admission PodEvictionAdmissi
 
 	result := []*apiv1.Pod{}
 	for _, podPrio := range calc.pods {
-		if admission == nil || admission.Admit(podPrio.pod, podPrio.recommendation) {
```
In the [same changeset](https://github.com/kubernetes/autoscaler/pull/5599/commits/4497aea552cc338e832dd5f389aae625eb545ac9) [there's a change to ensure that `admission` is always initialized and never `nil`](https://github.com/kubernetes/autoscaler/pull/5599/commits/4497aea552cc338e832dd5f389aae625eb545ac9).

[It is this `Updater` property that's passed into `GetSortedPods`](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/updater/logic/updater.go#L260), so I thought it should be safe to remove the nilcheck.

voelzmo commented on vertical-pod-autoscaler/e2e/v1/admission_controller.go:
```diff
@@ -19,6 +19,7 @@ package autoscaling
 import (
 	"context"
 	"fmt"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/test"
```
done.

voelzmo commented on vertical-pod-autoscaler/e2e/v1/recommender.go:
```diff
@@ -19,6 +19,7 @@ package autoscaling
 import (
 	"context"
 	"fmt"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/test"
```
done.

voelzmo commented on vertical-pod-autoscaler/e2e/v1/updater.go:
```diff
@@ -19,6 +19,7 @@ package autoscaling
 import (
 	"context"
 	"fmt"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/test"
```
done.

jbartosik commented:
LGTM but needs rebase

voelzmo commented:
@jbartosik rebased on master, please take another look. Unfortunately, the bot removes LGTM labels on new changes ;(

voelzmo commented:
@jbartosik can we merge this before the next rebase is necessary? In kind of afraid having to go through a lot of files again when other changes get merged 😱
Thanks!

jbartosik reviewed (approved):
/lgtm
/approve
